### ISSUE TITLE
Large scale training

### ISSUE THREAD

**Issue by iejMac** [lucidrains/magvit2-pytorch#17]
Hey, wanted to start this comm channel as I'm looking to do a large scale training run using some of this code. I'm happy to share graphs/samples as I go along and wanted to ask a few things to start off:

- Is the implementation of the original paper functionality complete?
- What is the best configuration you have found? (I'm seeing talks about LFQ vs FSQ and I see code for diff transformers etc.)

As always, thanks for this! Very helpful

**Comment by lucidrains** [comment]
@iejMac oh hey! yea, should be complete, probably one or two more bugs left to iron out

i would stick with LFQ initially, as that was what the magvit2 paper proposed, although some people have reported better results with FSQ. i put it into one repository so we can test them against each other and find out

**Comment by jpfeil** [comment]
Hi @iejMac I'd like to follow along if that's okay. It would be great if you could share any changes you make to the codebase to allow for larger scale training. I'm happy to share any weights I generate to help people get started with pretrained models.

**Comment by iejMac** [comment]
@jpfeil will do! 

Ok I think I'm mostly set up (had to port this code to a repo with a different style). My first question is - do we have some prepared configs (like what layers, how many frames, what fps etc.) which roughly correspond to some models they trained in the paper? Just so we cna compare.

**Comment by iejMac** [comment]
Oh I also noticed one thing - is there a reason we don't normalize the pixels before passing it into the model? Or did I just not catch where that's done?

**Comment by lucidrains** [comment]
@iejMac oh hey, what is the typical normalization for video? i think `.ToTensor()` [here](https://github.com/lucidrains/magvit2-pytorch/blob/main/magvit2_pytorch/data.py#L101) should bring it to `[0, 1]`?

**Comment by lucidrains** [comment]
@iejMac are you using the LFQ from [this repo](https://github.com/lucidrains/vector-quantize-pytorch#lookup-free-quantization)? the main claim of this paper is that this new quantization method helps them scale to more codes and better generation scores

**Comment by iejMac** [comment]
@lucidrains ah yeah ToTensor does but your VideoDataset doesn't do that and thats what I was using to test (was getting loss ~O(1e5)).
https://github.com/lucidrains/magvit2-pytorch/blob/b2f105b9a398264c523c2088f20b7b468200ab78/magvit2_pytorch/data.py#L159

**Comment by iejMac** [comment]
Yes I'm using LFQ from that. The main question I have about config is like can we figure out a parametrization of VideoTokenizer (given all params you added) that corresponds to like MAGVIT2-small so we can do some nice test runs.

Let's start out with - 8 frame videos at 25 FPS. Given that what are reasonable params for layers and other values in order to get decent results.

With the setup I sent above the loss curve/reconstructions look like this:
<img width="601" alt="Screenshot 2023-11-26 at 3 45 41 PM" src="https://github.com/lucidrains/magvit2-pytorch/assets/61431446/f81f7753-04af-4537-950d-4b0d35da8eb4">

<img width="281" alt="Screenshot 2023-11-26 at 3 45 59 PM" src="https://github.com/lucidrains/magvit2-pytorch/assets/61431446/38e63560-4fba-41cb-a8bb-71cb6c11857a">

**Comment by lucidrains** [comment]
@iejMac shoot, i normalized for gifs, but not mp4s.. thank you Maciej!

**Comment by lucidrains** [comment]
@iejMac yup, i can get some of the hyperparameters inline with the paper's probably Tuesday (currently in the middle of another project)

**Comment by iejMac** [comment]
cool, was just wondering if you have something on hand. I'll try to read/play around and I'll report here if I come up with something. Also for lowish-effort video dataloading video2numpy could be a good option! It's pretty fast and does all the normal preprocessing for you. Maybe I'll make a PR for that if you're interested.

**Comment by lucidrains** [comment]
@iejMac would greatly appreciate it! 🙏

**Comment by lucidrains** [comment]
@mudtriangle there's a BGR format? 😅

**Comment by mudtriangle** [comment]
Yep, and I think it's still default in cv2.

**Comment by lucidrains** [comment]
@mudtriangle got it, put in a quick fix [here](https://github.com/lucidrains/magvit2-pytorch/commit/286d4de7ce966ea7751114291ace10f4fbbe7528) (just doing it in tensor space, as i'm not familiar with cv2 enough)

**Comment by Jason3900** [comment]
> @mudtriangle got it, put in a quick fix [here](https://github.com/lucidrains/magvit2-pytorch/commit/286d4de7ce966ea7751114291ace10f4fbbe7528) (just doing it in tensor space, as i'm not familiar with cv2 enough)

Hello, I'm wondering if there's any progress of hyperparameter/architecture config alignment with the magvit-v2 paper.
